regex-assembly/include/932130.ra

```diff
@@ -1,16 +1,16 @@
 ##! Please refer to the documentation at
 ##! https://coreruleset.org/docs/development/regex_assembly/.
 
-\$\(.*\)
-\$\{.*\}
-<\(.*\)
->\(.*\)
+\$\([^()]*\)
+\$\{[^{}]*\}
+<\([^()]*\)
+>\([^()]*\)
```

Comment:
I think these patterns enable evasions. For example, `echo $(echo $(cat /etc/passwd))` would evade detection by `\$\([^\(\)]*\)`, because it uses parenthese inside of the subshell.  However, I think it works if you changed the expression to `\$\([^\)]*\)`, at a slight risk for more FPs. The match would now be `$(echo $(cat /etc/passwd)`, excluding the last parenthesis, but that is acceptable, IMO.
I would handle the other expressions in the same way.

In addition, I think it would be safe to use `+` instead of `*`. We wouldn't detect things like `>()`, but nothing meaningful can be done with simply invoking a subshell or creating an empty expansion (except when they are used for evasion , e.g. in `c${}a${}t /etc/passwd`, and I think we have those covered by other rules).

Reply:
Thanks. What I'm still trying to get is a regex that has less red flags in the ReDOS checker also.

Reply:
@theseion in the regex `\$\([^\(\)]*\)` the set [^\(\)] would match anything but `()`, so there is no subshell i.e `$()`. But now that you mention it, there could be brace expansion with {} and also ${}.

Reply:
Looks like fzipi planned both the scenarios.